### ISSUE TITLE
CKEditor5-link uses hardcoded strings such as 'link' and 'linkHref': prevents extension for 'Anchor' plugin

### ISSUE THREAD

**Issue by ConnorSinnott** [ckeditor/ckeditor5#5691]
## 📝 Provide a description of the new feature
My company requires an 'Anchor' plugin which, when used, would add an 'id' attribute to the selected element. We could then build a table of contents for the resulting HTML which would allow the user to jump to anchored elements. CKEditor4 had this feature previously.

I would say that the code to build the 'Anchor' plugin would be nearly identical to the code used in CKEditor5-link, save for the command and attribute names, and UI strings. The only functional difference between the plugins would be:

```JavaScript
// https://github.com/ckeditor/ckeditor5-link/blob/master/src/utils.js

/**
 * Creates link {@link module:engine/view/attributeelement~AttributeElement} with the provided `href` attribute.
 *
 * @param {String} href
 * @returns {module:engine/view/attributeelement~AttributeElement}
 */
export function createLinkElement( href, writer ) {
	// Priority 5 - https://github.com/ckeditor/ckeditor5-link/issues/121.
	const linkElement = writer.createAttributeElement( 'a', { href }, { priority: 5 } );
	writer.setCustomProperty( 'link', true, linkElement );

	return linkElement;
}

// From the above code
const linkElement = writer.createAttributeElement( 'a', { href }, { priority: 5 } );

// Would become
const anchorElement = writer.createAttributeElement( 'a', { id: name }, { priority: 5 } );
```

Unfortunately the Link plugin cannot be extended because while some strings are obtained through get methods which could be overridden
```JavaScript
/**
* @inheritDoc
 */
static get pluginName() {
    return 'Link';
}
```

Link specific magic strings such as 'link' and 'linkHref' are scattered throughout the files.

```JavaScript
/**
 * @inheritDoc
 */
refresh() {
	const model = this.editor.model;
	const doc = model.document;

	this.value = doc.selection.getAttribute( 'linkHref' );

	for ( const manualDecorator of this.manualDecorators ) {
		manualDecorator.value = this._getDecoratorStateFromModel( manualDecorator.id );
	}

	this.isEnabled = model.schema.checkAttributeInSelection( doc.selection, 'linkHref' );
}
```
With the code in its current state, my only option is to clone the link repository and make these changes manually.

Was this hard coding intentional? Could these strings be extracted into class methods? I would be happy to create a pull request, I just don't want to conflict with a code of practice. 

If you'd like to see this feature implemented, add a 👍 reaction to this post.

**Comment by Reinmar** [comment]
Hi! Interesting question. 

Code reusability is one of our goals in the core. It's getting less critical, though, when getting further towards end-user features.

Why is it so?

I agree that the anchor feature is quite similar to the link feature. Similar markup, perhaps a similar UI (although, with different labels). We could go further and say that, generally speaking, linking is like tagging with some specific attribute. Here we tag a selected text with `a[href]`. Somewhere else it could be `abbr[title]`, or `span[data-whatever]`. In all such cases, the UI will be quite similar too. 

So, we could start designing an abstraction that'd allow creating such features easily. We could end up with a base class with overridable getters for things that should differ or some fancy factory with dozens of params. 

Fast forward 3 years, and it turns out that our tagging feature generator became bloated with options and, thus, conditions. With time it turned out that, even though those specific features look similar to each other, they also bring a lot of custom behaviors here and there. E.g. the link feature has link decorators, it may have some validation in the UI, custom handling of inputted value, etc. Anchors should handle empty anchors (anchors with no text – just like in CKE4) and should validate their ids so there are no duplications. 

Alternatively, we'd stop extending our tagging feature generator, but that'd mean that with new requirements appearing with time, we'd move to custom implementations anyway.

tl;dr – the wrong abstraction is worse than code duplication. And here, correct abstraction is, from my experience, beyond our reach. In fact, I don't think it exists.

Instead of trying to generalize such features like link, we try to extract longer pieces of meaningful logic and make those reusable. Things that are quite complex and you'd rather like to have them in one place. Or because they are long and do completely repetitive things.

If there's any function in the ckeditor5-link package that could be extracted to a reasonable helper, we'd be for that. So far, the only such function was https://github.com/ckeditor/ckeditor5-link/blob/master/src/findlinkrange.js, but perhaps there's more. 

Let me know if this makes any sense :)

**Comment by mlewand** [comment]
Hi, It's been a while since we last heard from you. We're closing this issue for now. Still, feel free provide us requested feedback, so that we can reopen this issue.
